File: test/performance/data-structures/queue/deque.test.mjs
import { Deque } from '../../../../dist/esm/index.mjs';
import Benchmark from 'benchmark';
import { magnitude } from '../../../utils/index.mjs';

const suite = new Benchmark.Suite();
const { MILLION, HUNDRED_THOUSAND } = magnitude;

suite.add('1M push', function() {
  const deque = new Deque();
  for (let i = 0; i < MILLION; i++) {
    deque.push(i);
  }
  this.val = deque;
});

suite.add('Native JS Array 1M push', function() {
  const array = [];
  for (let i = 0; i < MILLION; i++) {
    array.push(i);
  }
  this.val = array;
});

suite.add('1M push & pop', function() {
  const deque = new Deque();
  for (let i = 0; i < MILLION; i++) {
    deque.push(i);
  }
  for (let i = 0; i < MILLION; i++) {
    deque.pop();
  }
  this.val = deque;
});

suite.add('Native JS Array 1M push & pop', function() {
  const array = [];
  for (let i = 0; i < MILLION; i++) {
    array.push(i);
  }
  for (let i = 0; i < MILLION; i++) {
    array.pop();
  }
  this.val = array;
});

suite.add('1M push & shift', function() {
  const deque = new Deque();
  for (let i = 0; i < MILLION; i++) {
    deque.push(i);
  }
  for (let i = 0; i < MILLION; i++) {
    deque.shift();
  }
  this.val = deque;
});

suite.add('100K push & shift', function() {
  const deque = new Deque();
  for (let i = 0; i < HUNDRED_THOUSAND; i++) {
    deque.push(i);
  }
  for (let i = 0; i < HUNDRED_THOUSAND; i++) {
    deque.shift();
  }
  this.val = deque;
});

suite.add('Native JS Array 100K push & shift', function() {
  const array = [];
  for (let i = 0; i < HUNDRED_THOUSAND; i++) {
    array.push(i);
  }
  for (let i = 0; i < HUNDRED_THOUSAND; i++) {
    array.shift();
  }
  this.val = array;
});

suite.add('100K unshift & shift', function() {
  const deque = new Deque();
  for (let i = 0; i < HUNDRED_THOUSAND; i++) {
    deque.unshift(i);
  }
  for (let i = 0; i < HUNDRED_THOUSAND; i++) {
    deque.shift();
  }
  this.val = deque;
});

suite.add('Native JS Array 100K unshift & shift', function() {
  const array = [];
  for (let i = 0; i < HUNDRED_THOUSAND; i++) {
    array.unshift(i);
  }
  for (let i = 0; i < HUNDRED_THOUSAND; i++) {
    array.shift();
  }
  this.val = array;
});

export { suite };

File: test/performance/data-structures/queue/deque.test.cpp
#include <benchmark/benchmark.h>
#include <deque>

const int MILLION = 1000000;
const int HUNDRED_THOUSAND = 100000;

std::string formatNumber(int num) {
  std::string result = std::to_string(num);
  int n = result.length() - 3;
  while (n > 0) {
    result.insert(n, ",");
    n -= 3;
  }
  return result;
}

// Test 1: 1M push
static void Push1M(benchmark::State& state) {
  for (auto _ : state) {
    std::deque<int> deque;
    for (int i = 0; i < MILLION; i++) {
      deque.push_back(i);
    }
    benchmark::DoNotOptimize(deque);
  }
}
BENCHMARK(Push1M)->Name("1M push");

// Test 2: 1M push & pop (LIFO - Stack behavior)
static void PushAndPop1M(benchmark::State& state) {
  for (auto _ : state) {
    std::deque<int> deque;
    for (int i = 0; i < MILLION; i++) {
      deque.push_back(i);
    }
    while (!deque.empty()) {
      deque.pop_back();
    }
    benchmark::DoNotOptimize(deque);
  }
}
BENCHMARK(PushAndPop1M)->Name("1M push & pop");

// Test 3: 1M push & shift (FIFO - Queue behavior)
static void PushAndShift1M(benchmark::State& state) {
  for (auto _ : state) {
    std::deque<int> deque;
    for (int i = 0; i < MILLION; i++) {
      deque.push_back(i);
    }
    while (!deque.empty()) {
      deque.pop_front();
    }
    benchmark::DoNotOptimize(deque);
  }
}
BENCHMARK(PushAndShift1M)->Name("1M push & shift");

// Test 4: 100K push & shift
static void PushAndShift100K(benchmark::State& state) {
  for (auto _ : state) {
    std::deque<int> deque;
    for (int i = 0; i < HUNDRED_THOUSAND; i++) {
      deque.push_back(i);
    }
    while (!deque.empty()) {
      deque.pop_front();
    }
    benchmark::DoNotOptimize(deque);
  }
}
BENCHMARK(PushAndShift100K)->Name("100K push & shift");

// Test 5: Native vector 100K push & shift (for comparison)
static void VectorPushAndShift100K(benchmark::State& state) {
  for (auto _ : state) {
    std::vector<int> arr;
    for (int i = 0; i < HUNDRED_THOUSAND; i++) {
      arr.push_back(i);
    }
    while (!arr.empty()) {
      arr.erase(arr.begin());
    }
    benchmark::DoNotOptimize(arr);
  }
}
BENCHMARK(VectorPushAndShift100K)->Name("Native vector 100K push & shift");

// Test 6: 100K unshift & shift (add to front)
static void UnshiftAndShift100K(benchmark::State& state) {
  for (auto _ : state) {
    std::deque<int> deque;
    for (int i = 0; i < HUNDRED_THOUSAND; i++) {
      deque.push_front(i);
    }
    while (!deque.empty()) {
      deque.pop_front();
    }
    benchmark::DoNotOptimize(deque);
  }
}
BENCHMARK(UnshiftAndShift100K)->Name("100K unshift & shift");

// Test 7: Native vector 100K unshift & shift (for comparison)
static void VectorUnshiftAndShift100K(benchmark::State& state) {
  for (auto _ : state) {
    std::vector<int> arr;
    for (int i = 0; i < HUNDRED_THOUSAND; i++) {
      arr.insert(arr.begin(), i);
    }
    while (!arr.empty()) {
      arr.erase(arr.begin());
    }
    benchmark::DoNotOptimize(arr);
  }
}
BENCHMARK(VectorUnshiftAndShift100K)->Name("Native vector 100K unshift & shift");

BENCHMARK_MAIN();
